File: dora/job/server/src/main/java/alluxio/dora/worker/JobWorker.java
/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

package alluxio.dora.worker;

import alluxio.dora.ClientContext;
import alluxio.dora.Constants;
import alluxio.dora.Server;
import alluxio.dora.client.file.FileSystem;
import alluxio.dora.client.file.FileSystemContext;
import alluxio.dora.conf.Configuration;
import alluxio.dora.conf.PropertyKey;
import alluxio.dora.exception.ConnectionFailedException;
import alluxio.dora.grpc.GrpcService;
import alluxio.dora.grpc.ServiceType;
import alluxio.dora.heartbeat.HeartbeatContext;
import alluxio.dora.heartbeat.HeartbeatThread;
import alluxio.dora.job.JobServerContext;
import alluxio.dora.metrics.MetricsSystem;
import alluxio.dora.user.ServerUserState;
import alluxio.dora.underfs.UfsManager;
import alluxio.dora.util.executor.ExecutorServiceFactories;
import alluxio.dora.wire.WorkerNetAddress;
import alluxio.dora.worker.job.command.CommandHandlingExecutor;
import alluxio.dora.worker.job.task.TaskExecutorManager;
import alluxio.dora.worker.job.JobMasterClient;
import alluxio.dora.job.JobMasterClientContext;

import com.google.common.base.Throwables;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.Future;
import javax.annotation.concurrent.NotThreadSafe;

/**
 * A job worker that manages all the worker-related activities.
 */
@NotThreadSafe
public final class JobWorker extends AbstractWorker {
  private static final Logger LOG = LoggerFactory.getLogger(JobWorker.class);

  private final JobServerContext mJobServerContext;
  /** Client for job master communication. */
  private final JobMasterClient mJobMasterClient;
  /** The service that handles commands sent from master. */
  private Future<?> mCommandHandlingService;

  /**
   * Creates a new instance of {@link JobWorker}.
   *
   * @param ufsManager the ufs manager
   */
  JobWorker(FileSystem filesystem, FileSystemContext fsContext, UfsManager ufsManager) {
    super(ExecutorServiceFactories.fixedThreadPool("job-worker-executor", 1));
    mJobServerContext = new JobServerContext(filesystem, fsContext, ufsManager);
    mJobMasterClient = JobMasterClient.Factory.create(JobMasterClientContext
        .newBuilder(ClientContext.create(Configuration.global())).build());
  }

  @Override
  public Set<Class<? extends Server>> getDependencies() {
    return new HashSet<>();
  }

  @Override
  public String getName() {
    return Constants.JOB_WORKER_NAME;
  }

  @Override
  public Map<ServiceType, GrpcService> getServices() {
    return Collections.emptyMap();
  }

  @Override
  public void start(WorkerNetAddress address) throws IOException {
    super.start(address);

    // Start serving metrics system, this will not block
    MetricsSystem.startSinks(Configuration.getString(PropertyKey.METRICS_CONF_FILE));

    try {
      JobWorkerIdRegistry.registerWorker(mJobMasterClient, address);
    } catch (ConnectionFailedException e) {
      LOG.error("Failed to connect to job master", e);
      throw Throwables.propagate(e);
    }
    TaskExecutorManager taskExecutorManager =
        new TaskExecutorManager(Configuration.getInt(PropertyKey.JOB_WORKER_THREADPOOL_SIZE),
            address);

    mCommandHandlingService = getExecutorService().submit(
        new HeartbeatThread(HeartbeatContext.JOB_WORKER_COMMAND_HANDLING,
            new CommandHandlingExecutor(mJobServerContext, taskExecutorManager, mJobMasterClient,
                address),
            (int) Configuration.getMs(PropertyKey.JOB_MASTER_WORKER_HEARTBEAT_INTERVAL),
            Configuration.global(), ServerUserState.global()));
  }

  @Override
  public void stop() throws IOException {
    if (mCommandHandlingService != null) {
      mCommandHandlingService.cancel(true);
    }
    mJobMasterClient.close();

    super.stop();
  }
}

File: cli/src/alluxio.org/cli/processes/job_worker.go
/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

package processes

import (
	"fmt"
	"strings"

	"github.com/palantir/stacktrace"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"

	"alluxio.org/cli/env"
)

var JobWorker = &WorkerProcess{
	BaseProcess: &env.BaseProcess{
		Name:                 "job_worker",
		JavaClassName:        "alluxio.worker.AlluxioJobWorker",
		JavaOpts:             ConfAlluxioJobWorkerJavaOpts,
		ProcessOutFile:       "job_worker.out",
		MonitorJavaClassName: "alluxio.worker.job.AlluxioJobWorkerMonitor",
	},
}

const (
	envAlluxioJobWorkerLogger = "ALLUXIO_JOB_WORKER_LOGGER"
	jobWorkerLoggerType       = "JOB_WORKER_LOGGER"
)

var (
	ConfAlluxioJobWorkerJavaOpts = env.RegisterTemplateEnvVar(&env.AlluxioConfigEnvVar{
		EnvVar:     "ALLUXIO_JOB_WORKER_JAVA_OPTS",
		IsJavaOpts: true,
	})
	confAlluxioJobWorkerAttachOpts = env.RegisterTemplateEnvVar(&env.AlluxioConfigEnvVar{
		EnvVar:     "ALLUXIO_JOB_WORKER_ATTACH_OPTS",
		IsJavaOpts: true,
	})
)

type JobWorkerProcess struct {
	*env.BaseProcess
}

func (p *JobWorkerProcess) Base() *env.BaseProcess {
	return p.BaseProcess
}

func (p *JobWorkerProcess) SetEnvVars(envVar *viper.Viper) {
	envVar.SetDefault(envAlluxioJobWorkerLogger, jobWorkerLoggerType)
	// ALLUXIO_JOB_WORKER_JAVA_OPTS = {default logger opts} ${ALLUXIO_JAVA_OPTS} ${ALLUXIO_WORKER_JAVA_OPTS}
	javaOpts := []string{
		fmt.Sprintf(env.JavaOptFormat, env.ConfAlluxioLoggerType, envVar.Get(envAlluxioJobWorkerLogger)),
	}
	javaOpts = append(javaOpts, env.ConfAlluxioJavaOpts.JavaOptsToArgs(envVar)...)
	javaOpts = append(javaOpts, p.JavaOpts.JavaOptsToArgs(envVar)...)
	envVar.Set(p.JavaOpts.EnvVar, strings.Join(javaOpts, " "))
}

func (p *JobWorkerProcess) StartCmd(cmd *cobra.Command) *cobra.Command {
	cmd.Use = p.Name
	return cmd
}

func (p *JobWorkerProcess) Start(cmd *env.StartProcessCommand) error {
	cmdArgs := []string{env.Env.EnvVar.GetString(env.ConfJava.EnvVar)}
	cmdArgs = append(cmdArgs, confAlluxioJobWorkerAttachOpts.JavaOptsToArgs(env.Env.EnvVar)...)
	cmdArgs = append(cmdArgs, "-cp", env.Env.EnvVar.GetString(env.EnvAlluxioServerClasspath))
	cmdArgs = append(cmdArgs, p.JavaOpts.JavaOptsToArgs(env.Env.EnvVar)...)
	cmdArgs = append(cmdArgs, p.JavaClassName)

	if err := p.Launch(cmd, cmdArgs); err != nil {
		return stacktrace.Propagate(err, "error launching process")
	}
	return nil
}

func (p *JobWorkerProcess) StopCmd(cmd *cobra.Command) *cobra.Command {
	cmd.Use = p.Name
	return cmd
}
